### PR TITLE
Mm 12447 Migrate 'components/get_link_modal.jsx' and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by Phizzard**
#### Summary
Migrate 'components/get_link_modal.jsx' and associated tests to TypeScript

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12447

### CODE REVIEW

**Comment by hanzei** [comment]
@Phizzard you have some linter errors:
```
/root/mattermost-webapp/components/get_link_modal.tsx
  15:18  error  Trailing spaces not allowed                                 no-trailing-spaces
  40:5   error  Missing accessibility modifier on class property copyLink   @typescript-eslint/explicit-member-accessibility
  43:9   error  Expected space(s) after "if"                                keyword-spacing
  43:21  error  Missing space before opening brace                          space-before-blocks
  56:13  error  'helpText' is never reassigned. Use 'const' instead         prefer-const
  62:10  error  Operator ':' must be spaced                                 space-infix-ops
  64:13  error  'copyLink' is never reassigned. Use 'const' instead         prefer-const
  77:10  error  Operator ':' must be spaced                                 space-infix-ops
  90:13  error  'copyLinkConfirm' is never reassigned. Use 'const' instead  prefer-const
  98:10  error  Operator ':' must be spaced                                 space-infix-ops

✖ 10 problems (10 errors, 0 warnings)
  9 errors and 0 warnings potentially fixable with the `--fix` option.
 ```

File: components/get_link_modal.tsx [hunk 1]
```diff
@@ -1,104 +1,101 @@
...
 
-    static defaultProps = {
+export default class GetLinkModal extends React.PureComponent<Props, State> {
+    private textAreaRef = React.createRef<HTMLTextAreaElement>();
+    public static defaultProps = {
         helpText: null,
     };
 
-    constructor(props) {
+    public constructor(props: Props) {
```
Comment (reflog) [inline components/get_link_modal.tsx]: :Props is not needed here

File: components/get_link_modal.tsx [hunk 2]
```diff
@@ -1,104 +1,101 @@
...
+                this.setState({copiedLink: document.execCommand('copy')});
+            } catch (err) {
+                this.setState({copiedLink: false});
+            }
         }
+    }
 
-        let copyLink = null;
+    public render(): JSX.Element {
+        let helpText: JSX.Element = this.props.helpText ? (
```
Comment (reflog) [inline components/get_link_modal.tsx]: don't think `: JSX.Element` is needed, it's inferred

File: components/get_link_modal.tsx [hunk 3]
```diff
@@ -1,104 +1,101 @@
...
-                <p className='alert alert-success alert--confirm'>
-                    <SuccessIcon/>
-                    <FormattedMessage
-                        id='get_link.clipboard'
-                        defaultMessage=' Link copied'
-                    />
-                </p>
-            );
-        }
+        let copyLinkConfirm: JSX.Element = this.state.copiedLink ? (
```
Comment (reflog) [inline components/get_link_modal.tsx]: why the change in code? now this renders an empty div

File: components/get_link_modal.tsx [hunk 4]
```diff
@@ -1,104 +1,101 @@
...
-                    onClick={this.copyLink}
-                >
-                    <FormattedMessage
-                        id='get_link.copy'
-                        defaultMessage='Copy Link'
-                    />
-                </button>
-            );
-        }
+        let copyLink: JSX.Element = document.queryCommandSupported('copy') ? (
```
Comment (reflog) [inline components/get_link_modal.tsx]: why the change in code? now this renders an empty div

File: components/get_link_modal.tsx (hunk 3 above)
Comment (Phizzard) [reply components/get_link_modal.tsx]: The code change was for the JSX.Element. Short-circuiting caused `copyLinkConfirm` to potentially be undefined which was not allowed with the JSX.Element.

However, based on your previous comment above JSX.Element type can be inferred so I can change this back to before the change.

File: components/get_link_modal.tsx (hunk 1 above)
Comment (Phizzard) [reply components/get_link_modal.tsx]: without `:Props` check-types runs into the following error
```
Parameter 'props' implicitly has an 'any' type.

28     public constructor(props) {
```

**Comment by Phizzard** [comment]
Anyone know how to get around the current type-check error I'm getting in the test file using setProps on the wrapper returned from `shallowIntl` helper?

```
components/get_link_modal.test.tsx:50:27 - error TS2345: Argument of type '{ show: boolean; }' is not assignable to parameter of type 'Pick<{ intl: InjectedIntl; }, "intl">'.
  Object literal may only specify known properties, and 'show' does not exist in type 'Pick<{ intl: InjectedIntl; }, "intl">'.

50         wrapper.setProps({show: true});
```

I can't seem to find anything how to get around it

File: package.json [hunk 5]
```diff
@@ -11,6 +11,7 @@
     "@types/jest": "24.0.17",
     "@types/node": "12.7.0",
     "@types/react": "16.8.24",
+    "@types/react-bootstrap": "^0.32.20",
```
Comment (saturninoabril) [inline package.json]: Please remove `^` as normally goes with fix version.

File: components/get_link_modal.tsx (hunk 1 above)
Comment (mickmister) [reply components/get_link_modal.tsx]: > I also don't understand why it can't infer type of props, since it's clearly stated in type signature of React.PureComponent

I believe this is because the declaration of props in the signature does not define what this class's constructor's parameters are, but rather the parent class `PureComponent` (with the generic type included). Also, it defines what `props` this component takes, but I think that is slightly different than the constructor's parameters on a class/object level, even though we tend to think of them and treat them as one in the same.

As for using `noImplicitAny`, having the flag off makes the compiler too lenient and leads to incorrect or missed typings. Code review needs to be much more strict to catch the missed typings, whereas a strict compiler does much of the work for us. It's meant to discourage you from explicitly using `: any` or `as any` in your code, as that is a sign that your types could be defined/used better.

Comment (reflog) [reply components/get_link_modal.tsx]: Well put. Let's leave it as is. :+1:

File: components/get_link_modal.tsx [hunk 6]
```diff
@@ -1,104 +1,101 @@
...
-                <p className='alert alert-success alert--confirm'>
-                    <SuccessIcon/>
-                    <FormattedMessage
-                        id='get_link.clipboard'
-                        defaultMessage=' Link copied'
-                    />
-                </p>
-            );
-        }
+        const copyLinkConfirm = this.state.copiedLink && (
```
Comment (saturninoabril) [inline components/get_link_modal.tsx]: Is this change necessary?  If not, please revert to original form.  Same case for `copyLink` and `helpText` above.

**Comment by saturninoabril** [comment]
@Phizzard Please see comment - https://github.com/mattermost/mattermost-webapp/pull/3954#discussion_r335886548

**Comment by saturninoabril** [comment]
@Phizzard Let me know status of https://github.com/mattermost/mattermost-webapp/pull/3954#discussion_r335886548.  I can help push changes per request if you don't mind.

File: components/get_link_modal.tsx (hunk 6 above)
Comment (saturninoabril) [reply components/get_link_modal.tsx]: @Phizzard If you don't mind, I'll push commit per requested change.
